.test-infra/mock-apis/src/main/go/internal/service/echo/echo.go

```diff
@@ -0,0 +1,224 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package echo
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/cache"
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/logging"
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/metric"
+	echov1 "github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/proto/echo/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	metricsNamePrefix = "echo"
+	echoPath          = "/proto.echo.v1.EchoService/Echo"
+	echoPathAlias     = "/v1/echo"
+	healthPath        = "/grpc.health.v1.Health/Check"
+	healthPathAlias   = "/v1/healthz"
+)
+
+var (
+	defaultLogger = logging.New("github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/service/echo")
+)
+
+// Option applies optional parameters to the echo service.
+type Option interface {
+	apply(srv *echo)
+}
+
+// WithLogger overrides the default logging.Logger.
+func WithLogger(logger logging.Logger) Option {
+	return &loggerOpt{
+		v: logger,
+	}
+}
+
+// WithLoggerFields WitLoggerFields supplies the echo service's logging.Logger with logging.Field slice.
+func WithLoggerFields(fields ...logging.Field) Option {
+	return &loggerFieldsOpts{
+		v: fields,
+	}
+}
+
+// WithMetricWriter supplies the echo service with a metric.Writer.
+func WithMetricWriter(writer metric.Writer) Option {
+	return &metricWriterOpt{
+		v: writer,
+	}
+}
+
+// Register a grpc.Server with the echov1.EchoService. Returns a http.Handler or error.
+func Register(s *grpc.Server, decrementer cache.Decrementer, opts ...Option) (http.Handler, error) {
+	srv := &echo{
+		decrementer: decrementer,
+		logger:      defaultLogger,
+	}
+	for _, opt := range opts {
+		opt.apply(srv)
+	}
+
+	echov1.RegisterEchoServiceServer(s, srv)
+	grpc_health_v1.RegisterHealthServer(s, srv)
+
+	return srv, nil
+}
+
+type echo struct {
+	echov1.UnimplementedEchoServiceServer
+	grpc_health_v1.UnimplementedHealthServer
+	decrementer   cache.Decrementer
+	logger        logging.Logger
+	loggerFields  []logging.Field
+	metricsWriter metric.Writer
+}
+
+// ServeHTTP implements http.Handler, allowing echo to support HTTP clients in addition to gRPC.
+func (srv *echo) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if f, ok := map[string]http.HandlerFunc{
+		echoPath:        srv.httpHandler,
+		echoPathAlias:   srv.httpHandler,
+		healthPath:      srv.checkHandler,
+		healthPathAlias: srv.checkHandler,
+	}[r.URL.Path]; ok {
+		f(w, r)
+		return
+	}
+	http.Error(w, fmt.Sprintf("%s not found", r.URL.Path), http.StatusNotFound)
```

Comment:
```suggestion
	switch r.URL.Path {
	case echoPath, echoPathAlias:
  		srv.httpHandler(w, r)
	case healthPath, healthPathAlias:
		srv.checkHandler(w,r)
	default:
		http.Error(w, fmt.Sprintf("%s not found", r.URL.Path), http.StatusNotFound)
	}
```

Using a switch statement clarifies that some of the cases are identical, and avoids the "indent error path" reversal.